# Table of contents

## Importing libraries
## Importing data
## Grouping data
## Aggregating variables

# Import libraries

In [2]:
# import libraries
import pandas as pd
import numpy as np
import os

# Answer 1


In [3]:
path = r'C:\Users\Mafalda\Documents\mafalda\Formação\Data Analytics\CareerFoundry\Data Analytics Program\2.Data Immersion\Achievement 4\Instacart Basket Analysis'

In [4]:
df_ords_prods_merge = pd.read_pickle(os.path.join(path,'02 Data', 'Prepared Data','orders_products_merged_v2.pkl'))

In [6]:
df_ords_prods_merge.shape

(32434212, 19)

In [22]:
df_ords_prods_merge.columns

Index(['order_id', 'user_id', 'eval_set', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', '_merge', 'product_name', 'aisle_id',
       'department_id', 'prices', 'Price_range_loc', 'busiest_day',
       'busiest_days', 'busiest_period_of_day', 'max_order', 'loyalty_flag'],
      dtype='object')

# Answer 2

In [8]:
# Grouping dataframe by product_ name column
df_ords_prods_merge.groupby('product_name')

In [9]:
# Aggregating data wirh agg()
df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457687
2                17.277920
3                17.179756
4                17.811403
5                15.213779
6                16.439806
7                17.225773
8                15.340520
9                15.895474
10               20.197148
11               16.170828
12               15.887622
13               16.583304
14               16.757377
15               16.165037
16               17.663250
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

In [10]:
# Checking the same infoo without agg()
df_ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457687
2     17.277920
3     17.179756
4     17.811403
5     15.213779
6     16.439806
7     17.225773
8     15.340520
9     15.895474
10    20.197148
11    16.170828
12    15.887622
13    16.583304
14    16.757377
15    16.165037
16    17.663250
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

In [11]:
# Checking mean, max and min on the order_number column
df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                15.457687   1  99
2                17.277920   1  99
3                17.179756   1  99
4                17.811403   1  99
5                15.213779   1  99
6                16.439806   1  99
7                17.225773   1  99
8                15.340520   1  99
9                15.895474   1  99
10               20.197148   1  99
11               16.170828   1  99
12               15.887622   1  99
13               16.583304   1  99
14               16.757377   1  99
15               16.165037   1  99
16               17.663250   1  99
17               15.694469   1  99
18               19.310397   1  99
19               17.177343   1  99
20               16.473447   1  99
21               22.902379   1  99

# Answer 3

###### in the subset it was identified ( "department Id"_16 - mean 19,463012; department Id"_17 - mean 11.294069) as the highest and lowest mean values. In the total dataframe ( "department Id"_21 - mean 22.902379; department Id"_5 - mean 15.213779), as expected since it holds all the data records.

# Answer 4

In [12]:
# Creating loyalty flag column for exsiting customers usinf transform() and loc() functions
df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [13]:
df_ords_prods_merge.head(15)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
5    3367565        1    prior             6                   2   
6     550135        1    prior             7                   1   
7    3108588        1    prior             8                   1   
8    2295261        1    prior             9                   1   
9    2550362        1    prior            10                   4   
10   2968173       15    prior            15                   1   
11   1870022       15    prior            17                   2   
12   1911383       15    prior            18                   2   
13   2715276       15    prior            21                   1   
14    487368       15    prior            22                   1   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
5                   7                    19.0         196                  1   
6                   9                    20.0         196                  1   
7                  14                    14.0         196                  2   
8                  16                     0.0         196                  4   
9                   8                    30.0         196                  1   
10                  9                     7.0         196                  2   
11                 16                     8.0         196                  6   
12                 11                     7.0         196                  1   
13                  9                     7.0         196                  1   
14                 10                    14.0         196                  1   

    reordered _merge product_name  aisle_id  department_id  prices  \
0           0   both         Soda        77              7     9.0   
1           1   both         Soda        77              7     9.0   
2           1   both         Soda        77              7     9.0   
3           1   both         Soda        77              7     9.0   
4           1   both         Soda        77              7     9.0   
5           1   both         Soda        77              7     9.0   
6           1   both         Soda        77              7     9.0   
7           1   both         Soda        77              7     9.0   
8           1   both         Soda        77              7     9.0   
9           1   both         Soda        77              7     9.0   
10          0   both         Soda        77              7     9.0   
11          1   both         Soda        77              7     9.0   
12          1   both         Soda        77              7     9.0   
13          1   both         Soda        77              7     9.0   
14          1   both         Soda        77              7     9.0   

       Price_range_loc     busiest_day         busiest_days  \
0    Mid range product  Regularly busy       Regularly busy   
1    Mid range product  Regularly busy           Least busy   
2    Mid range product  Regularly busy           Least busy   
3    Mid range product      Least busy  Bottom busiest busy   
4    Mid range product      Least busy  Bottom busiest busy   
5    Mid range product  Regularly busy       Regularly busy   
6    Mid range product  Regularly busy   Second Busiest day   
7    Mid 

In [14]:
# Creating a flag
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [15]:
# Printing the frequency of the loyalty _flag column
df_ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

loyalty_flag
Regular customer    15891077
Loyal customer      10293737
New customer         6249398
Name: count, dtype: int64

In [16]:
df_ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

# Answer 5

In [23]:
# performing basic statistics on customer spending habits

df_ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.388747  1.0  99999.0
New customer      13.294370  1.0  99999.0
Regular customer  12.496203  1.0  99999.0

In [ ]:
# From the data we can see that there is not a big difference in the average price numbers between the three categories, but there is visible trend on mean price reduction among returning customers.

# Answer 6

In [30]:
# Creating a new column for "mean price" for product price range
df_ords_prods_merge['mean_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [28]:
df_ords_prods_merge. columns


Index(['order_id', 'user_id', 'eval_set', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', '_merge', 'product_name', 'aisle_id',
       'department_id', 'prices', 'Price_range_loc', 'busiest_day',
       'busiest_days', 'busiest_period_of_day', 'max_order', 'loyalty_flag',
       'mean price'],
      dtype='object')

In [31]:
# Creating a Spending flag 
df_ords_prods_merge.loc[df_ords_prods_merge['mean_price'] < 10, 'Spending_flag'] = 'Low spender'

In [33]:
df_ords_prods_merge.loc[df_ords_prods_merge['mean_price'] >= 10, 'Spending_flag'] = 'High spender'

In [34]:
# Printing the frequency of the Spending_flag column
df_ords_prods_merge['Spending_flag'].value_counts(dropna=False)

Spending_flag
Low Spender     31798751
High spender      635461
Name: count, dtype: int64

In [36]:
df_ords_prods_merge. head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered  ... prices     Price_range_loc     busiest_day  \
0          0  ...    9.0   Mid range product  Regularly busy   
1          1  ...    9.0   Mid range product  Regularly busy   
2          1  ...    9.0   Mid range product  Regularly busy   
3          1  ...    9.0   Mid range product      Least busy   
4          1  ...    9.0   Mid range product      Least busy   

          busiest_days  busiest_period_of_day max_order  loyalty_flag  \
0       Regularly busy          Fewest orders        10  New customer   
1           Least busy          Fewest orders        10  New customer   
2           Least busy             Most oders        10  New customer   
3  Bottom busiest busy          Fewest orders        10  New customer   
4  Bottom busiest busy             Most oders        10  New customer   

  mean price mean_price  Spending_flag  
0   6.367797   6.367797    Low Spender  
1   6.367797   6.367797    Low Spender  
2   6.367797   6.367797    Low Spender  
3   6.367797   6.367797    Low Spender  
4   6.367797   6.367797    Low Spender  

[5 rows x 24 columns]

# Answer 7

In [39]:
# Creating a median " days_since_prior_order" column
df_ords_prods_merge['median_days_since_prior_order'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [40]:
df_ords_prods_merge. head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered  ...     Price_range_loc     busiest_day         busiest_days  \
0          0  ...   Mid range product  Regularly busy       Regularly busy   
1          1  ...   Mid range product  Regularly busy           Least busy   
2          1  ...   Mid range product  Regularly busy           Least busy   
3          1  ...   Mid range product      Least busy  Bottom busiest busy   
4          1  ...   Mid range product      Least busy  Bottom busiest busy   

   busiest_period_of_day  max_order  loyalty_flag mean price mean_price  \
0          Fewest orders         10  New customer   6.367797   6.367797   
1          Fewest orders         10  New customer   6.367797   6.367797   
2             Most oders         10  New customer   6.367797   6.367797   
3          Fewest orders         10  New customer   6.367797   6.367797   
4             Most oders         10  New customer   6.367797   6.367797   

  Spending_flag  median_days_since_prior_order  
0   Low Spender                           20.5  
1   Low Spender                           20.5  
2   Low Spender                           20.5  
3   Low Spender                           20.5  
4   Low Spender                           20.5  

[5 rows x 25 columns]

In [52]:
# Creating a Frequency_flag column
df_ords_prods_merge.loc[df_ords_prods_merge['median_days_since_prior_order'] > 20, 'Frequency_flag'] = 'Non-frequent customer"'

In [53]:
df_ords_prods_merge.loc[(df_ords_prods_merge['median_days_since_prior_order'] > 10) & (df_ords_prods_merge['median_days_since_prior_order'] <= 20) , 'Frequency_flag'] = 'Regular customer"'

In [54]:
df_ords_prods_merge.loc[df_ords_prods_merge['median_days_since_prior_order'] <= 10, 'Frequency_flag'] = 'Frequent customer"'

In [58]:
# Printing the frequency of the Frequency_flag column
df_ords_prods_merge['Frequency_flag'].value_counts(dropna=False)

Frequency_flag
Frequent customer"        21577409
Regular customer"          7217134
Non-frequent customer"     3639669
Name: count, dtype: int64

In [56]:
df_ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered  ...     busiest_day         busiest_days  busiest_period_of_day  \
0          0  ...  Regularly busy       Regularly busy          Fewest orders   
1          1  ...  Regularly busy           Least busy          Fewest orders   
2          1  ...  Regularly busy           Least busy             Most oders   
3          1  ...      Least busy  Bottom busiest busy          Fewest orders   
4          1  ...      Least busy  Bottom busiest busy             Most oders   

   max_order  loyalty_flag mean price mean_price Spending_flag  \
0         10  New customer   6.367797   6.367797   Low Spender   
1         10  New customer   6.367797   6.367797   Low Spender   
2         10  New customer   6.367797   6.367797   Low Spender   
3         10  New customer   6.367797   6.367797   Low Spender   
4         10  New customer   6.367797   6.367797   Low Spender   

  median_days_since_prior_order          Frequency_flag  
0                          20.5  Non-frequent customer"  
1                          20.5  Non-frequent customer"  
2                          20.5  Non-frequent customer"  
3                          20.5  Non-frequent customer"  
4                          20.5  Non-frequent customer"  

[5 rows x 26 columns]

In [61]:
df_ords_prods_merge.shape

(32434212, 26)

# Answer 9

In [62]:
# Exporting new dataframe as a pickle file
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_v3.pkl'))